### PR TITLE
Add FSharpWebSite plus simple functional test against it

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
Note that the initial goal here is **not** to provide comprehensive test coverage of MVC sites written with F#. It's simply to provide a structure where we can insert any F#-specific tests that we want in the future. I will shortly be adding a test covering actions that return `FSharpAsync<T>` (once the corresponding PR is merged in the `Common` repo).

Hopefully, having F# code in the repo will not cause any trouble for people working on MVC who don't have the FSharp tools. The F# website here builds and runs without needing anything extra installed (the F# SDK comes via NuGet), and I haven't added `FSharpWebSite.fsproj` to the VS `.sln` files.

### CODE REVIEW

**Comment by rynowak** [comment]
> I haven't added FSharpWebSite.fsproj to the VS .sln files.

I think we either need to add it to Mvc.sln OR find a different home for this (maybe aspnet/Entropy). I think it would be nice if we had a dedicated integration tests repo but Entropy is currently the best home for something like this.

File: test/WebSites/FSharpWebSite/FSharpWebSite.fsproj [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    <Compile Include="Controllers\*.fs" />
+    <Compile Include="Startup.fs" />
+    <Compile Include="Program.fs" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\src\Microsoft.AspNetCore.Mvc\Microsoft.AspNetCore.Mvc.csproj" />
+
+    <PackageReference Include="FSharp.Core" Version="$(FSharpCoreVersion)" />
+    <PackageReference Include="FSharp.NET.Sdk" Version="1.0.*" PrivateAssets="All" />
```
Comment (rynowak) [inline test/WebSites/FSharpWebSite/FSharpWebSite.fsproj]: Should this have its own version in `dependencies.props`?

**Comment by rynowak** [comment]
I think this all looks straightforward, we just need to understand where it should live. I'm not opposed to adding it to the MVC repo **if** we are willing to add it to MVC.sln. I don't want any projects that aren't in *some* solution. 

What's the experience like if you don't have the F# tools? Does it just nag you once about not being able to load the project? I think now that 2017 is the baseline that's not a big deal.

/cc @Eilon

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/FSharpWebSiteTest.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
+﻿using System.Net;
```
Comment (Eilon) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/FSharpWebSiteTest.cs]: Add license header

File: test/WebSites/FSharpWebSite/Controllers/HomeController.fs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+﻿namespace FSharpWebSite.Controllers
```
Comment (Eilon) [inline test/WebSites/FSharpWebSite/Controllers/HomeController.fs]: License header.

File: test/WebSites/FSharpWebSite/FSharpWebSite.fsproj [hunk 4]
```diff
@@ -0,0 +1,38 @@
+﻿<Project Sdk="FSharp.NET.Sdk; Microsoft.NET.Sdk.Web">
+  <Import Project="..\..\..\build\dependencies.props" />
+
+  <PropertyGroup>
+    <TargetFrameworks>net46;netcoreapp2.0</TargetFrameworks>
```
Comment (Eilon) [inline test/WebSites/FSharpWebSite/FSharpWebSite.fsproj]: Remove `net46` from here and use `<TargetFramework>` (singular) tag instead. See my email from yesterday for info.

File: test/WebSites/FSharpWebSite/FSharpWebSite.fsproj [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    <TargetFrameworks Condition="'$(OS)' != 'Windows_NT'">netcoreapp2.0</TargetFrameworks>
```
Comment (Eilon) [inline test/WebSites/FSharpWebSite/FSharpWebSite.fsproj]: And I think you can omit this line entirely.

File: test/WebSites/FSharpWebSite/Controllers/HomeController.fs (hunk 3 above)
Comment (Eilon) [inline test/WebSites/FSharpWebSite/Controllers/HomeController.fs]: (And all other new test files)

**Comment by rynowak** [comment]
> What's the experience like if you don't have the F# tools? Does it just nag you once about not being able to load the project? I think now that 2017 is the baseline that's not a big deal.

So I tried this out on the 15.1 version of the F# tools and it is missing this `FSharp.NET.Sdk` SDK. So, this project file might depend on 15.3 features.

**Comment by SteveSandersonMS** [comment]
@rynowak:

> What's the experience like if you don't have the F# tools?

There is no VS experience for F# + .NET Core right now, whether or not you have the F# tools, as per [this announcement post](https://blogs.msdn.microsoft.com/dotnet/2017/03/07/announcing-f-4-1-and-the-visual-f-tools-for-visual-studio-2017-2/):

> However, the following are not supported at this time, but will be supported in a future Visual Studio 2017 update:
>    * F# support for .NET Core Visual Studio tooling

So it doesn't really help us to put the `.fsproj` into a `.sln` file right now, whether that's in `Mvc.sln` or elsewhere. If we do so, the project still won't appear in VS, whether you have the F# tooling or not.

**Suggestion:** How about we don't put it in an `.sln` right now, but instead leave an open issue assigned to me to add it once the relevant F# tooling is released? Hopefully it can go into `Mvc.sln` then (that would be the most useful place, so you can implement and test any MVC F#-specific functionality atomically), but we might choose to migrate it out to Entropy if it turns out that the VS2017.3 experience is problematic if F# tools aren't installed.

File: test/WebSites/FSharpWebSite/FSharpWebSite.fsproj (hunk 1 above)
Comment (SteveSandersonMS) [inline test/WebSites/FSharpWebSite/FSharpWebSite.fsproj]: Done

File: test/WebSites/FSharpWebSite/FSharpWebSite.fsproj (hunk 4 above)
Comment (SteveSandersonMS) [inline test/WebSites/FSharpWebSite/FSharpWebSite.fsproj]: Done

File: test/WebSites/FSharpWebSite/FSharpWebSite.fsproj (hunk 5 above)
Comment (SteveSandersonMS) [inline test/WebSites/FSharpWebSite/FSharpWebSite.fsproj]: Done

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/FSharpWebSiteTest.cs (hunk 2 above)
Comment (SteveSandersonMS) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/FSharpWebSiteTest.cs]: Done

**Comment by rynowak** [comment]
Yeah that seems like an OK resolution. It would be good to get an understanding of when that's going to land.

File: test/WebSites/FSharpWebSite/Controllers/HomeController.fs (hunk 3 above)
Comment (SteveSandersonMS) [inline test/WebSites/FSharpWebSite/Controllers/HomeController.fs]: Done

File: test/WebSites/FSharpWebSite/Controllers/HomeController.fs [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+open System.Collections.Generic
+open System.Linq
+open System.Threading.Tasks
+open Microsoft.AspNetCore.Mvc
+
+type HomeController () =
+    inherit Controller()
+
+    member this.Index () =
+        this.View()
```
Comment (panesofglass) [inline test/WebSites/FSharpWebSite/Controllers/HomeController.fs]: Add the another member to test F# Async (along with the accompanying view):

``` fsharp
    member this.Sleepy () =
        async {
            // Simulate a call that takes two seconds.
            do! Async.Sleep 2000
            return this.View()
        }
```

**Comment by SteveSandersonMS** [comment]
@Eilon:

> Remove net46 from here and use <TargetFramework> (singular) tag instead

I had to revert that change and put `net46` back. Omitting it was breaking the build, because upstream projects (functional tests) still include that target framework. We can remove this later as part of removing it from everywhere else.

**Comment by SteveSandersonMS** [comment]
@rynowak OK, have filed https://github.com/aspnet/Mvc/issues/6238

@Eilon @rynowak Any other CR points, or is this good to merge?

**Comment by SteveSandersonMS** [comment]
@Eilon Oh, putting `net46` back is also breaking the build! Maybe the upstream code has changed by now. I'll revert the revert...

**Comment by SteveSandersonMS** [comment]
Thanks!

File: test/WebSites/FSharpWebSite/Controllers/HomeController.fs (hunk 6 above)
Comment (SteveSandersonMS) [inline test/WebSites/FSharpWebSite/Controllers/HomeController.fs]: @panesofglass Thanks - will add this once the F# support in the Common repo is built and becomes available in the MVC repo.

**Pull request merged**